### ISSUE TITLE
Bug: Behavior of onFocus event in React 17 is counterintuitive

### ISSUE THREAD

**Issue by devongovett** [facebook/react#19743]
React version: 17.0.0-rc.1

## Steps To Reproduce

https://codesandbox.io/s/blue-waterfall-pvwss?file=/src/App.js:0-599

1. Click on the input element
2. Notice that only "focusin on div" is logged, not "focus on input"

## The current behavior

In React 17 RC, `onFocus` on a child element is never called if a parent element calls `stopPropagation` on the native event first. In React 16 `onFocus` was fired as expected. This affected the way the [FocusScope](https://github.com/adobe/react-spectrum/blob/main/packages/@react-aria/focus/src/FocusScope.tsx#L231-L271) component is implemented in React Aria, which uses native `focusin` event handlers and relies on browser bubbling behavior.

I believe this is happening because React is now listening using a non-capturing `focusin` listener on the React root rather than a capturing `focus` listener on the document body. Calling `stopPropagation` on the div in this example prevents the `focusin` event from reaching the React root, and thus the `onFocus` prop on the child input is never called.

I understand that this is due to the way React synthetic events propagate vs. native events, but it is quite counterintuitive that stopping propagation on a parent element would prevent a child from receiving an event so I thought I'd call it out before the stable 17 release.

## The expected behavior

`onFocus` should be fired on child elements if a parent calls stopPropagation. Perhaps this could be fixed by using a capturing listener at the react root level rather than a non-capturing one? This way React's synthetic events would run before native events bubble. Another way to fix it would be to register event handlers directly on elements rather than relying on event delegation, but I imagine that would be a much bigger change.

We cannot switch to capturing listeners in React Aria because we don't want to prevent propagation from going down the tree, only upward to parent FocusScopes. The problem for us is that the React root is above the current scope, so other focus handlers on child elements within a scope won't receive focus events anymore.

**Comment by eps1lon** [comment]
This is indeed confusing behavior.

Though React 17 is a major release and breaking changes are expected. It'd be nice to explain the concrete use case. Ideally a test that was passing in React 16 and is failing in 17 (with all the relevant source code inlined). Then it's easier to discuss possible alternatives in case the current behavior for React 17 can't be changed.

**Comment by gaearon** [comment]
The intent was to make the behavior consistent — this is how it already worked for most other events. For example, clicks. Why should focus behave differently in that sense?

Making all bubble listeners fire in bubble phase and capture listeners fire in capture phase was also an intentional decision so that (1) there's fewer differences between React events and native events and (2) so that it is possible to more safely nest React trees. In particular, the behavior you're describing is what would allow the inner tree's `stopPropagation` call to avoid reaching the outer tree.

**Comment by devongovett** [comment]
That's fair. To be clear, I think it's counterintuitive for all events, not just focus, but this is the one that happened to fail our tests.

I was trying to think of an alternative way of implementing this so it would behave as expected, but I didn't come up with anything. If you switched to capturing listeners, then you'd have the same problem in reverse. `stopPropagation` on a parent wouldn't stop it from reaching a child.

At this point, what is the benefit of adding all events to the root and implementing custom bubbling rather than adding events directly to the elements the event handler prop is on? I imagine at one time it was performance, but is that still the case in modern browsers? Is whatever performance benefit still exists worth the counterintuitive behavior it creates?

React always has a reference to the element being rendered, so it's not needing to traverse the DOM tree to find elements to bind, which was the original motivation for event delegation in general. Can users who need to avoid whatever perf cost there is to `addEventListener` do so themselves by adding their own listener to a parent element and relying on bubbling? This would be much more explicit about where events are bound.

If it's event bubbling through portals (which I'd also argue is counterintuitive), then perhaps that could be implemented by re-dispatching the event in the DOM on the portal parent? Or maybe in this case, actually implementing manual bubbling above the portal root (to preserve `event.target`)?

I think Jason Miller's [post](https://jasonformat.com/event-delegation-vs-direct-binding/) on event delegation vs direct binding recently was a good read. It seems to me that a major version of React is a perfect time to re-evaluate this. Perhaps you've already done so, and in that case, I'd love to know the reasoning.

**Comment by devongovett** [comment]
Ah, perhaps before hydration root level events could be used to record, but switch to direct event handlers afterward?

For concurrent mode, are you talking about adding/removing listeners from the DOM replacing the handler before it should be? Maybe react could have a single handler attached, but swap out the user's handler internally?

```js
// On initial render.
let onClick = props.onClick;
node.addEventListener('click', (e) => {
  onClick(e);
});

// sometime later, e.g. in next render or when new tree is flushed
onClick = newProps.onClick;
```

I'm probably misunderstanding though, and I know you would have already thought all of this through. I'm mostly wondering whether the tradeoff is worth it, or whether there's a way to get the best of both worlds somehow.

**Comment by gaearon** [comment]
I think we're going to proceed with the 17 semantics here — although we might reevaluate in a future release. Are you hard-blocked on this problem

**Comment by devongovett** [comment]
I believe we managed to work around this for now (cc. @snowystinger).

**Comment by gaearon** [comment]
OK, I'm going to close as working as intended.
